clouddrift/ragged.py

```diff
@@ -794,6 +831,12 @@ def _mask_var(
     elif isinstance(criterion, (list, np.ndarray, xr.DataArray)):
         # select multiple values
         mask = np.isin(var, criterion)
+    elif callable(
+        criterion
+    ):  # mask directly created by applying `criterion` to each trajectory
+        mask = xr.DataArray(
+            data=apply_ragged(criterion, var, rowsize), dims=[dim_name]
```

Comment:
Last comment, this will eventually crash if you pass a function that does not returns one boolean per element. Apply ragged would still work, but that would cause an issue with `np.logical_and` up there...

I can't think of a good way to catch this beforehand, but maybe you do?

Reply:
You're right. Actually, maybe in this case it crashes right away because of `dims` being specified?
One way to crash nicely could be to test the size of `apply_ragged` returned value. But I don't think it is possible to make sure beforehand that `apply_ragged` result will be compatible.

Reply:
I guess that would work, I don't see how we can do it without executing `apply_ragged`. I think that would be good to throw an exception if the returned `mask` doesn't have the correct dimensions. But thinking about this I tried another example and modify the test.

```
func = lambda arr: arr <= 2  # keep id smaller or equal to 2
ds_sub = subset(self.ds, {"id": func})
self.assertTrue(all(ds_sub["id"] == [1, 2]))
self.assertTrue(all(ds_sub["rowsize"] == [5, 4]))
```

and it doesn't work because we are always passing `rowsize` to ragged array. It has to work for both `obs` and `traj` variables.

Reply:
See, the latest commit. It might be possible to improve but now it works for all variables.

Reply:
I was focused on the `obs` case indeed... Your latest commit seems fine to me!

Reply:
I'm a bit late to the party, I was travelling yesterday and had a really quick look at the commit.
Maybe I can propose a small re-write (for consistency):
```python
        if len(var) == len(rowsize):
            mask = criterion(var)
        else:
            mask = apply_ragged(criterion, var, rowsize)

        mask = xr.DataArray(data=mask, dims=[dim_name]).astype(bool)
```